reframe/core/schedulers/slurm.py

```diff
@@ -162,63 +155,104 @@ def submit(self):
 
         self._jobid = int(jobid_match.group('jobid'))
 
-    def _count_compatible_nodes(self, nodes):
-        constraints = set()
-        partitions = (set(self.sched_partition.split())
-                      if self.sched_partition else set())
-        excluded_node_names = self._get_excluded_node_names()
-
-        if self.options:
-            for optstr in self.options:
-                optstr = optstr.strip()
-                if optstr.startswith('--'):
-                    optstr = optstr.replace('=', ' ', 1)
-
-                option, arg = optstr.split(maxsplit=1)
-                if option == '-C' or option == '--constraint':
-                    constraints.update(arg.split())
-
-                if option == '-p' or option == '--partition':
-                    partitions.update(arg.split())
-
-        num_nodes = 0
-        for n in nodes:
-            if (n.active_features >= constraints and
-                n.partitions >= partitions and
-                n.name not in excluded_node_names):
-                num_nodes += 1
-
-        return num_nodes
-
-    def _get_reservation_nodes(self):
-        command = 'scontrol show res %s' % self.sched_reservation
+    def _show_nodes(self):
+        command = 'scontrol show -o nodes'
+        completed = os_ext.run_command(command, check=True)
+        if completed:
+            node_descriptions = completed.stdout.splitlines()
+        else:
+            raise JobError('could not show the nodes')
+
+        return {SlurmNode(descr) for descr in node_descriptions}
+
+    def get_available_nodes(self):
+        nodes = self._show_nodes()
+        option_parser = ArgumentParser()
+        option_parser.add_argument('-C', '--constraint')
```

Comment:
Applied the suggested change

Reply:
This function is restrictive, because it only allows virtual partitions to be specified with constraints and/or partitions. This is wrong. Any combination of options that produces a set of nodes may be used for a virtual partition. So practically, this function should do the same thing as the `filter_nodes`, but on the options passed to the `sched_access` variable.